clippy_lints/src/functions.rs

```diff
@@ -268,6 +390,138 @@ impl<'a, 'tcx> Functions {
+                Applicability::MachineApplicable,
+            );
+        }
+    });
+}
+
+fn returns_unit(decl: &hir::FnDecl) -> bool {
+    match decl.output {
+        hir::FunctionRetTy::DefaultReturn(_) => true,
+        hir::FunctionRetTy::Return(ref ty) => match ty.node {
+            hir::TyKind::Tup(ref t) => t.is_empty(),
+            hir::TyKind::Never => true,
+            _ => false,
+        },
+    }
+}
+
+fn must_use_span(attrs: &[Attribute]) -> Option<Span> {
+    attrs.iter().find_map(|attr| {
+        attr.ident().and_then(|ident| {
+            if "must_use" == &ident.as_str() {
+                Some(attr.span)
+            } else {
+                None
+            }
+        })
+    })
+}
+
+fn has_mutable_arg(cx: &LateContext<'_, '_>, body: &hir::Body) -> bool {
+    let mut tys = FxHashSet::default();
+    body.params.iter().any(|param| is_mutable_pat(cx, &param.pat, &mut tys))
+}
+
+fn is_mutable_pat(cx: &LateContext<'_, '_>, pat: &hir::Pat, tys: &mut FxHashSet<DefId>) -> bool {
+    if let hir::PatKind::Wild = pat.node {
+        return false; // ignore `_` patterns
+    }
+    let def_id = pat.hir_id.owner_def_id();
+    if cx.tcx.has_typeck_tables(def_id) {
+        is_mutable_ty(cx, &cx.tcx.typeck_tables_of(def_id).pat_ty(pat), tys)
+    } else {
+        false
+    }
+}
+
+static KNOWN_IMMUTABLES: &[&[&str]] = &[&["core", "marker", "PhantomData"]];
+
+static KNOWN_MUTABLES: &[&[&str]] = &[
+    &["core", "sync", "atomic", "AtomicBool"],
+    &["core", "sync", "atomic", "AtomicI8"],
+    &["core", "sync", "atomic", "AtomicI16"],
+    &["core", "sync", "atomic", "AtomicI32"],
+    &["core", "sync", "atomic", "AtomicI64"],
+    &["core", "sync", "atomic", "AtomicI128"],
+    &["core", "sync", "atomic", "AtomicIsize"],
+    &["core", "sync", "atomic", "AtomicU8"],
+    &["core", "sync", "atomic", "AtomicU16"],
+    &["core", "sync", "atomic", "AtomicU32"],
+    &["core", "sync", "atomic", "AtomicU64"],
+    &["core", "sync", "atomic", "AtomicU128"],
+    &["core", "sync", "atomic", "AtomicUsize"],
+    &["core", "cell", "UnsafeCell"],
+    &["core", "cell", "Cell"],
+    &["core", "cell", "RefCell"],
+    &["std", "sync", "Mutex"],
+    &["std", "sync", "RwLock"],
+    &["std", "sync", "mpsc", "Sender"],
+    &["std", "sync", "mpsc", "Receiver"], //TODO: Does recv count here?
+];
+
+static KNOWN_WRAPPER_TYS: &[&[&str]] = &[&["core", "rc", "Rc"], &["core", "rc", "Arc"]];
+
+fn is_mutable_ty(cx: &LateContext<'_, '_>, ty: Ty<'_>, tys: &mut FxHashSet<DefId>) -> bool {
+    use ty::TyKind::*;
+    match ty.sty {
+        // primitive types are never mutable
+        Bool | Char | Int(_) | Uint(_) | Float(_) | Str => false,
+        Adt(ref adt, ref substs) => {
+            if !tys.insert(adt.did) {
+                return false;
+            }
+            !KNOWN_IMMUTABLES.iter().any(|path| match_def_path(cx, adt.did, path))
+                && KNOWN_MUTABLES.iter().any(|path| match_def_path(cx, adt.did, path))
```

Comment:
Can [`TyS::is_freeze`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyS.html#method.is_freeze) be used here?

Reply:
Great idea! That catches many more instances than my hacky logic. Need to fix the fallout, though. :smiley: